xapian-applications/omega/clickmodel/simplifieddbn.cc

```diff
@@ -148,50 +150,48 @@ SimplifiedDBN::buildSessions(string logfile)
 void
 SimplifiedDBN::train(vector<vector<string>> &sessions)
 {
-    map<string, map<string, map<char, array<double, 2>>>> urlRelFractions;
+    map<string, map<string, map<char, array<double, 2>>>> url_rel_fractions;
 
     for (size_t i = 0; i < sessions.size(); ++i) {
 	string qid = sessions[i][QID];
 
-	vector<string> docids = getDocidsList(sessions);
+	vector<string> docids = get_docid_list(sessions);
 
-	vector<int> clicks = getClicksList(sessions);
+	vector<int> clicks = get_click_list(sessions);
 
-	int lastClickedPos = clicks.size() - 1;
+	int last_clicked_pos = clicks.size() - 1;
 
 	for (size_t j = 0; j < clicks.size(); ++j)
 	    if (clicks[j] != 0)
-		lastClickedPos = j;
+		last_clicked_pos = j;
```

Comment:
If the final entry in `clicks` is non-zero, then `last_clicked_pos` will be `clicks.size() - 1` (this is also the case if all entries in `clicks` are `0`).

But then in the loop below, on the final iteration `x` will be `last_clicked_pos + 1` which is `clicks.size()` and then `clicks[x]` results in undefined behaviour.

I think you want `x <= last_clicked_pos` below, which would copy over all clicks up to and including the final non-zero one.